src/main/java/de/tum/in/www1/artemis/service/QuizService.java

```diff
@@ -0,0 +1,274 @@
+                        }
+                    }
+                }
+                quizQuestionStatistic.getShortAnswerSpotCounters().removeAll(spotCounterToDelete);
+
+                // save references as index to prevent Hibernate Persistence problem
+                saveCorrectMappingsInIndicesShortAnswer(saQuestion);
+            }
+        }
+
+        T savedQuizConfiguration = saveAndFlush(quizConfiguration);
+
+        // fix references in all drag and drop questions and short answer questions (step 2/2)
+        for (QuizQuestion quizQuestion : savedQuizConfiguration.getQuizQuestions()) {
+            if (quizQuestion instanceof DragAndDropQuestion dragAndDropQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndices(dragAndDropQuestion);
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion shortAnswerQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesShortAnswer(shortAnswerQuestion);
+            }
+        }
+
+        return savedQuizConfiguration;
+    }
+
+    /**
+     * remove dragItem and dropLocation from correct mappings and set dragItemIndex and dropLocationIndex instead
+     *
+     * @param dragAndDropQuestion the question for which to perform these actions
+     */
+    private void saveCorrectMappingsInIndices(DragAndDropQuestion dragAndDropQuestion) {
+        List<DragAndDropMapping> mappingsToBeRemoved = new ArrayList<>();
+        for (DragAndDropMapping mapping : dragAndDropQuestion.getCorrectMappings()) {
+            // check for NullPointers
+            if (mapping.getDragItem() == null || mapping.getDropLocation() == null) {
+                mappingsToBeRemoved.add(mapping);
+                continue;
+            }
+
+            // drag item index
+            DragItem dragItem = mapping.getDragItem();
+            boolean dragItemFound = false;
+            for (DragItem questionDragItem : dragAndDropQuestion.getDragItems()) {
+                if (dragItem.equals(questionDragItem)) {
+                    dragItemFound = true;
+                    mapping.setDragItemIndex(dragAndDropQuestion.getDragItems().indexOf(questionDragItem));
+                    mapping.setDragItem(null);
+                    break;
+                }
+            }
+
+            // drop location index
+            DropLocation dropLocation = mapping.getDropLocation();
+            boolean dropLocationFound = false;
+            for (DropLocation questionDropLocation : dragAndDropQuestion.getDropLocations()) {
+                if (dropLocation.equals(questionDropLocation)) {
+                    dropLocationFound = true;
+                    mapping.setDropLocationIndex(dragAndDropQuestion.getDropLocations().indexOf(questionDropLocation));
+                    mapping.setDropLocation(null);
+                    break;
+                }
+            }
+
```

Comment:
Code duplication, refactor into a function with generic parameter and return a boolean if the item has been found.

Reply:
Code is refactored